File: 0238-product-of-array-except-self/solution.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int prefix = 1, suffix = 1, n = nums.size();
        //vector<int> res(n,1);
        int res[n]; // using raw array

        for (int i = 0; i < n; i++) {
            res[i] = prefix;
            prefix *= nums[i];
        }

        for (int i = n-1; i >= 0; i--) {
            res[i] = res[i] * suffix;
            suffix *= nums[i];
        }

        // convert back to vector
        return vector<int>(res, res+n);
    }
};

File: 0238-product-of-array-except-self/solution.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n, prefix, postfix = len(nums), 1, 1
        res = [1] * n

        # two loops, prefix and suffix.
        for i in range(n):
            res[i] = prefix
            prefix *= nums[i]
        
        for i in reversed(range(n)):
            res[i] *= postfix
            postfix *= nums[i]

        return res
